awss3/store.go

```diff
@@ -0,0 +1,703 @@
+		client    *s3.S3
+		sess      *session.Session
+		endpoint  string
+		bucket    string
+		cachepath string
+	}
+
+	object struct {
+		fs         *FS
+		o          *s3.GetObjectOutput
+		cachedcopy *os.File
+
+		name      string    // aka "key" in s3
+		updated   time.Time // LastModifyied in s3
+		metadata  map[string]string
+		bucket    string
+		readonly  bool
+		opened    bool
+		cachepath string
+
+		infoOnce sync.Once
+		infoErr  error
+	}
+)
+
+// NewClient create new AWS s3 Client.  Uses cloudstorage.Config to read
+// necessary config settings such as bucket, region, auth.
+func NewClient(conf *cloudstorage.Config) (*s3.S3, *session.Session, error) {
+
+	awsConf := aws.NewConfig().
+		WithHTTPClient(http.DefaultClient).
+		WithMaxRetries(aws.UseServiceDefaultRetries).
+		WithLogger(aws.NewDefaultLogger()).
+		WithLogLevel(aws.LogOff).
+		WithSleepDelay(time.Sleep)
+
+	if conf.Region != "" {
+		awsConf.WithRegion(conf.Region)
+	} else {
+		awsConf.WithRegion("us-east-1")
+	}
+
+	switch conf.AuthMethod {
+	case AuthAccessKey:
+		accessKey := conf.Settings.String(ConfKeyAccessKey)
+		if accessKey == "" {
+			return nil, nil, ErrNoAccessKey
+		}
+		secretKey := conf.Settings.String(ConfKeyAccessSecret)
+		if secretKey == "" {
+			return nil, nil, ErrNoAccessSecret
+		}
+		awsConf.WithCredentials(credentials.NewStaticCredentials(accessKey, secretKey, ""))
+	default:
+		return nil, nil, ErrNoAuth
+	}
+
+	if conf.BaseUrl != "" {
+		awsConf.WithEndpoint(conf.BaseUrl).WithS3ForcePathStyle(true)
+	}
+
+	disableSSL := conf.Settings.Bool(ConfKeyDisableSSL)
+	if disableSSL {
+		awsConf.WithDisableSSL(true)
+	}
+
+	sess := session.New(awsConf)
+	if sess == nil {
+		return nil, nil, ErrNoS3Session
+	}
+
+	s3Client := s3.New(sess)
+
+	return s3Client, sess, nil
+}
+
+// NewStore Create AWS S3 storage client of type cloudstorage.Store
+func NewStore(c *s3.S3, sess *session.Session, conf *cloudstorage.Config) (*FS, error) {
+
+	if conf.TmpDir == "" {
+		return nil, fmt.Errorf("unable to create cachepath. config.tmpdir=%q", conf.TmpDir)
+	}
+	err := os.MkdirAll(path.Dir(conf.TmpDir), 0775)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create cachepath. config.tmpdir=%q err=%v", conf.TmpDir, err)
+	}
+
+	uid := uuid.NewUUID().String()
+	uid = strings.Replace(uid, "-", "", -1)
+
+	return &FS{
+		client:    c,
+		sess:      sess,
+		bucket:    conf.Bucket,
+		cachepath: conf.TmpDir,
+		ID:        uid,
+		PageSize:  cloudstorage.MaxResults,
+	}, nil
+}
+
+// Type of store = "s3"
+func (f *FS) Type() string {
+	return StoreType
+}
+
+// Client gets access to the underlying google cloud storage client.
+func (f *FS) Client() interface{} {
+	return f.client
+}
+
+// String function to provide s3://..../file   path
+func (f *FS) String() string {
+	return fmt.Sprintf("s3://%s/", f.bucket)
+}
+
+// NewObject of Type s3.
+func (f *FS) NewObject(objectname string) (cloudstorage.Object, error) {
+	obj, err := f.Get(context.Background(), objectname)
```

Comment:
We need to create a ticket to add better `context` support, that is support contexts for all the methods..  Anyway, just a todo for one of us.

Reply:
i added quite a few more, but i agree, still some to go.  i might add this one.